lib/dal/src/billing_publish.rs

```diff
@@ -0,0 +1,174 @@
+//! This module provides the ability to publish billing events.
+
+#![warn(
+    bad_style,
+    clippy::missing_panics_doc,
+    clippy::panic,
+    clippy::panic_in_result_fn,
+    clippy::unwrap_in_result,
+    clippy::unwrap_used,
+    dead_code,
+    improper_ctypes,
+    missing_debug_implementations,
+    missing_docs,
+    no_mangle_generic_items,
+    non_shorthand_field_patterns,
+    overflowing_literals,
+    path_statements,
+    patterns_in_fns_without_body,
+    unconditional_recursion,
+    unreachable_pub,
+    unused,
+    unused_allocation,
+    unused_comparisons,
+    unused_parens,
+    while_true
+)]
+
+use billing_events::{
+    BillingEventsError, BillingWorkspaceChangeEvent, BillingWorkspaceChangeEventKind,
+};
+use telemetry::prelude::*;
+use thiserror::Error;
+
+use crate::{
+    resource_metadata::{ResourceMetadata, ResourceMetadataError},
+    ChangeSet, ChangeSetError, ComponentId, DalContext, TransactionsError,
+};
+
+#[allow(missing_docs)]
+#[remain::sorted]
+#[derive(Debug, Error)]
+pub enum BillingPublishError {
+    #[error("billing events error: {0}")]
+    BillingEvents(#[from] BillingEventsError),
+    #[error("change set error: {0}")]
+    ChangeSet(#[from] ChangeSetError),
+    #[error("resource metadata error: {0}")]
+    ResourceMetadata(#[from] ResourceMetadataError),
+    #[error("transactions error: {0}")]
+    Transactions(#[from] TransactionsError),
+}
+
+type BillingPublishResult<T> = Result<T, BillingPublishError>;
+
+#[instrument(
+    name = "billing_publish.for_resource_count",
+    level = "debug",
+    skip(ctx, change_set)
+)]
+pub(crate) async fn for_resource_count(
+    ctx: &DalContext,
+    change_set: &ChangeSet,
+) -> BillingPublishResult<()> {
+    if change_set.workspace_id.is_none() || !change_set.is_head(ctx).await? {
+        warn!(
+            ?change_set,
+            "cannot publish billing workspace change event for non-HEAD change set"
+        );
+        return Ok(());
+    };
+
+    let metadata = ResourceMetadata::list(ctx).await?;
+    let resource_count = metadata.len();
+
+    let workspace_id = change_set.workspace_id()?;
+    let event = BillingWorkspaceChangeEvent {
```

Comment:
Are we worried about bloat if we also send the `metadata` itself here?

Reply:
Not at all! I think it will be in there if we want it to be. Truthfully, I think I was going to see how big it was first. Worst case we can chunk it.